src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/BannedTypesAnalyzer.cs

```diff
@@ -0,0 +1,97 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Azure.ClientSdk.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class BannedTypesAnalyzer : SymbolAnalyzerBase
+    {
+        private static HashSet<string> BannedTypes = new HashSet<string>()
+        {
+            "Azure.Core.Json.MutableJsonDocument",
+            "Azure.Core.Json.MutableJsonElement",
+            "Azure.Core.Json.MutableJsonChange",
+            "Azure.Core.Json.MutableJsonChangeKind",
+        };
+
+        private static readonly string BannedTypesMessageArgs = string.Join(", ", BannedTypes);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Descriptors.AZC0020);
+
+        public override SymbolKind[] SymbolKinds { get; } = new[]
+        {
+            SymbolKind.Method,
+            SymbolKind.Field,
+            SymbolKind.Property,
+            SymbolKind.Parameter,
+            SymbolKind.Event,
+            SymbolKind.NamedType,
+        };
+
+        public override void Analyze(ISymbolAnalysisContext context)
+        {
+            if (IsAzureCore(context.Symbol.ContainingAssembly))
+            {
+                return;
+            }
+
+            switch (context.Symbol)
+            {
+                case IParameterSymbol parameterSymbol:
+                    CheckType(context, parameterSymbol.Type, parameterSymbol);
+                    break;
+                case IMethodSymbol methodSymbol:
+                    CheckType(context, methodSymbol.ReturnType, methodSymbol);
+                    break;
+                case IEventSymbol eventSymbol:
+                    CheckType(context, eventSymbol.Type, eventSymbol);
+                    break;
+                case IPropertySymbol propertySymbol:
+                    CheckType(context, propertySymbol.Type, propertySymbol);
+                    break;
+                case IFieldSymbol fieldSymbol:
+                    CheckType(context, fieldSymbol.Type, fieldSymbol);
+                    break;
```

Comment:
Do we have tests for these cases? I'd like to see what kind of code is about to be banned.

Reply:
I will add them per this comment: https://github.com/Azure/azure-sdk-tools/pull/6642#discussion_r1290303416